### PR TITLE
CB-15994 Enable patch upgrade for Ranger RAZ clusters for 7.2.7 or newer

### PR DESCRIPTION
**Pull request by daszabo**
See detailed description in the commit message.

### CODE REVIEW

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java [hunk 1]
```diff
@@ -84,29 +86,39 @@ public UpgradeV4Response checkForUpgrade(NameOrCrn nameOrCrn, Long workspaceId,
...
-                throw new BadRequestException(message);
+            String runtimeVersion = runtimeVersionService.getRuntimeVersion(stack.getCluster().getId()).orElse("");
+            boolean razGaSupport = isVersionNewerOrEqualThanLimited(runtimeVersion, CLOUDERA_STACK_VERSION_7_2_7);
+            if (razGaSupport && datalakeCluster.isRangerRazEnabled()) {
+                LOGGER.debug("DataHub patch upgrade is enabled for Ranger RAZ clusters with stack version 7.2.7 or newer.");
+                return filteredCandidates.stream()
+                        .filter(imgInfo -> runtimeVersion.equals(imgInfo.getComponentVersions().getCdp()))
+                        .collect(Collectors.toList());
+            } else {
+                boolean rangerRazEnabled = datalakeCluster.isRangerRazEnabled();
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java:107]: Could be moved before the if-else as it's used 2 times here.

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java [hunk 2]
```diff
@@ -84,29 +86,39 @@ public UpgradeV4Response checkForUpgrade(NameOrCrn nameOrCrn, Long workspaceId,
...
-                    "Data Hub Upgrade is %s as Ranger RAZ is %s for [%s] cluster.",
-                    rangerRazEnabled ? "not allowed" : "allowed",
-                    rangerRazEnabled ? "enabled" : "disabled",
-                    datalakeCluster.getName());
-            LOGGER.debug(message);
-            if (rangerRazEnabled) {
-                throw new BadRequestException(message);
+            String runtimeVersion = runtimeVersionService.getRuntimeVersion(stack.getCluster().getId()).orElse("");
+            boolean razGaSupport = isVersionNewerOrEqualThanLimited(runtimeVersion, CLOUDERA_STACK_VERSION_7_2_7);
+            if (razGaSupport && datalakeCluster.isRangerRazEnabled()) {
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java:101]: razGaSupport -> razPatchSupport

**Review by keyki — APPROVED** [review]

File: core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java [hunk 3]
```diff
@@ -84,29 +86,39 @@ public UpgradeV4Response checkForUpgrade(NameOrCrn nameOrCrn, Long workspaceId,
...
-                throw new BadRequestException(message);
+            if (razPatchSupport && rangerRazEnabled) {
+                LOGGER.debug("DataHub patch upgrade is enabled for Ranger RAZ clusters with stack version 7.2.7 or newer.");
+                return filteredCandidates.stream()
+                        .filter(imgInfo -> runtimeVersion.equals(imgInfo.getComponentVersions().getCdp()))
+                        .collect(Collectors.toList());
+            } else {
+                String message = String.format(
+                        "Data Hub Upgrade is %s as Ranger RAZ is %s for [%s] cluster.",
+                        rangerRazEnabled ? "not allowed" : "allowed",
```
Comment (keyki) [inline core/src/main/java/com/sequenceiq/distrox/v1/distrox/service/upgrade/DistroXUpgradeAvailabilityService.java:110]: Isn't `not allowed` and `allowed` swithced?

**Review by keyki — APPROVED** [review]

**Comment by daszabo** [comment]
retest this please

**Pull request merged**